### PR TITLE
[sdk] Hardcode known histograms using seconds

### LINKED ISSUE #4797: Update default values for explicit histogram bucket boundaries
# Feature Request

Currently the OpenTelemetry SDK sets default histogram boundaries to { 0, 5, 10, 25, 50, 75,
100, 250, 500, 750, 1000, 2500, 5000, 7500, 10000}. These values make sense for time durations of network
requests measured in ms, however OTel semantic conventions encourage using seconds. This means that most histograms will be timing things that take less than 5 seconds and the default buckets are not at all useful. I am in particular concerned with histograms that were added in .NET 8 to ASP.NET Core and System.Net because I expect it will be extremely common for users to want to monitor those histograms. The current workaround is to specify a custom view with alternate buckets for every histogram. This approach functions, but adds complexity to what should be a very simple and common use case. You can see how the getting started guide for ASP.NET Core metrics has to include a custom view right now: https://learn.microsoft.com/en-us/aspnet/core/log-mon/metrics/metrics?view=aspnetcore-8.0#create-the-starter-app

There are several tasks that could help improve the situation on different timescales:
- Hardcode the names of well-known instruments and provide more appropriate defaults for those
- Change the global histogram defaults for OpenTelemetry.NET
- Change the global histogram defaults in the specification for all OpenTelemetry clients
- Advance the metrics hint APIs towards stability and work with the .NET libraries team in .NET 9 to implement those APIs so that instrumentation authors can provide their own defaults.

The .NET team is glad to work with you on that last bullet in the .NET 9 timeframe, but we are hoping one of the other options can be adopted as well to offer a solution more quickly.

cc @JamesNK

### PR DESCRIPTION
**Pull request by samlam**
Closes #4797 
Design discussion issue #4797

## Changes

* Change histogram default boundaries when `MetricStreamIdentity.Unit` is set to "s"
* New default buckets are set to `{ 0, 0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1, 2.5, 5, 7.5, 10 }`
* Created a set of well-known Meters/Instruments for the default
* Added unit-test for the new defaults

## Acknowledgement

The PR is created with the assistance from @CodeBlanch

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [ ] Appropriate `[CHANGELOG.md](http://changelog.md/)` files updated for non-trivial changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry/Metrics/AggregatorStore.cs
Comment (reyang) [inline src/OpenTelemetry/Metrics/AggregatorStore.cs:27]: Could this be an IReadOnlyDictionary?

File: test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs
Comment (reyang) [inline test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs:242]: nit: consider alphabetical order

File: src/OpenTelemetry/Metrics/AggregatorStore.cs
Comment (utpilla) [inline src/OpenTelemetry/Metrics/AggregatorStore.cs:27]: Move this to `Metric` class (similar to how we have the histogram bounds information there)

Comment (utpilla) [inline src/OpenTelemetry/Metrics/AggregatorStore.cs:27]: All the entries have the same value. Use a `HashSet` instead?

File: test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs
Comment (utpilla) [inline test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs:254]: Add another unit test to check that a regular Histogram with `Unit` as seconds still uses the default Histogram bounds.

File: test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs
Comment (utpilla) [inline test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs:269]: This way of testing is relying a bit too much on the knowledge of the internal implementation. Could you use test the bounds using an `InMemoryExporter` instead? Get hold of the `MetricPoint` after the export and check the bounds using `metricPoint.GetHistogramBuckets().ExplicitBounds`.

Comment (samlam) [reply test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs:270]: I'm not familiar with InMemoryExporter, let me research it a bit. Thx

File: src/OpenTelemetry/Metrics/Metric.cs
Comment (vishweshbankwar) [inline src/OpenTelemetry/Metrics/Metric.cs:40]: `http.server.duration` and `http.client.duration` should not be part of this list. Those will continue to follow default buckets with unit as `ms`. https://github.com/open-telemetry/opentelemetry-dotnet/pull/4802

Comment (utpilla) [inline src/OpenTelemetry/Metrics/Metric.cs:40]:
```suggestion
        ("OpenTelemetry.Instrumentation.AspNetCore", "http.server.request.duration"),
```

File: src/OpenTelemetry/Metrics/Metric.cs
Comment (utpilla) [inline src/OpenTelemetry/Metrics/Metric.cs:41]:
```suggestion
        ("OpenTelemetry.Instrumentation.Http", "http.client.request.duration"),
```

File: test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs
Comment (utpilla) [reply test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs:255]: Is this resolved? I don't see another unit test added.

File: test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs
Comment (CodeBlanch) [reply test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs:270]: > This way of testing is relying a bit too much on the knowledge of the internal implementation.

This logic is strange to me. It's a unit test. We're changing how `AggregatorStore` resolves buckets. That's the "unit" we are testing. Logic downstream reliant on the resolved buckets isn't being changed. Other tests _should_ be covering that. I would very much expect unit tests to cover internal details 😄 

If we want a bigger integration test exercising more of the SDK, fine to do, but too much of that is going to result in slow CI with a ton of overlap. Just my $0.02 😄

Comment (utpilla) [reply test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs:270]: > That's the "unit" we are testing.

We haven't really followed the pedantic approach to unit testing in our repo. It's a spectrum. The tests in our repo are always somewhere between a strict by-the-book unit test and a blanket integration test.

File: src/OpenTelemetry/Metrics/Metric.cs
Comment (utpilla) [inline src/OpenTelemetry/Metrics/Metric.cs:32]: nit: Provide the capacity in the ctor

File: test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs
Comment (cijothomas) [reply test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs:270]: Please make sure to add tests to cover that this special logic is only applicable if SDK is using defaults. i.e if a View is configured to override buckets, then that gets respected in all metrics, including the ones we are hardcoding.

Comment (cijothomas) [reply test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs:270]: > Please make sure to add tests to cover that this special logic is only applicable if SDK is using defaults. i.e if a View is configured to override buckets, then that gets respected in all metrics, including the ones we are hardcoding.

And I recommend to the existing pattern of using InMemoryExporter to achieve this, though I fully agree its neither unit nor integration tests but something in between!

Comment (CodeBlanch) [reply test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs:270]: > Please make sure to add tests to cover that this special logic is only applicable if SDK is using defaults. i.e if a View is configured to override buckets, then that gets respected in all metrics, including the ones we are hardcoding.

Why is that needed? This code is only touching the default bounds which are used if there wasn't a view. We have view tests covering view bounds are selected over the defaults (whatever they may be).

@samlam You could try extending [this test](https://github.com/open-telemetry/opentelemetry-dotnet/blob/34544bdb5effa80eca3facd7438763764a8063be/test/OpenTelemetry.Tests/Metrics/MetricViewTests.cs#L513). But not sure how to best solve needing an exact meter name.

> We haven't really followed the pedantic approach to unit testing in our repo. It's a spectrum. The tests in our repo are always somewhere between a strict by-the-book unit test and a blanket integration test.

Well we are the maintainers. If we ask people to add integration tests that's what we'll get 😄 And we'll have to maintain the extra code forever!

Comment (utpilla) [reply test/OpenTelemetry.Tests/Metrics/AggregatorTestsBase.cs:270]: > This code is only touching the default bounds which are used if there wasn't a view.

We know this to be true but it would be good to test that and verify that Views take precedence even for especially treated metrics.

**Review by utpilla — APPROVED** [review]
Approving this as the changes made in the `src/` look good to me and we would like to unblock `1.6.0` release.

The unresolved comments should be addressed in later PRs.

**Review by CodeBlanch — APPROVED** [review]
LGTM thanks @samlam

File: src/OpenTelemetry/Metrics/Metric.cs
Comment (utpilla) [inline src/OpenTelemetry/Metrics/Metric.cs:42]: It would be great to update the PR description with the links to the runtime code for where these metrics are defined.

**Pull request merged**
